File: ghidra_scripts/PuntoHookReport.java
//@author Codex
//@category Punto

import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.program.model.listing.Program;
import ghidra.program.model.symbol.ExternalLocation;
import ghidra.program.model.symbol.ExternalLocationIterator;
import ghidra.program.model.symbol.ExternalManager;
import ghidra.util.exception.CancelledException;

import java.io.File;
import java.io.PrintWriter;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class PuntoHookReport extends GhidraScript {

    private static final Set<String> TOKENS = new HashSet<>(Arrays.asList(
        "hook", "switch", "layout", "caps", "diary", "password",
        "caret", "copy", "paste", "mouse", "escape", "f12"
    ));

    @Override
    protected void run() throws Exception {
        Program p = currentProgram;
        if (p == null) {
            println("No current program");
            return;
        }

        File outFile = new File(System.getProperty("user.dir"), "ghidra_pshook64_report.txt");
        try (PrintWriter out = new PrintWriter(outFile, "UTF-8")) {
            writeLine(out, "Program: " + p.getName());
            writeLine(out, "Language: " + p.getLanguage());
            writeLine(out, "");

            writeLine(out, "== Functions (interesting names) ==");
            FunctionIterator it = p.getFunctionManager().getFunctions(true);
            while (it.hasNext()) {
                Function fn = it.next();
                String name = fn.getName();
                if (matches(name)) {
                    writeLine(out, name + " @ " + fn.getEntryPoint());
                }
            }

            writeLine(out, "");
            writeLine(out, "== Imports (interesting) ==");
            ExternalManager em = p.getExternalManager();
            for (String lib : em.getExternalLibraryNames()) {
                writeLine(out, "[" + lib + "]");
                ExternalLocationIterator locIt = em.getExternalLocations(lib);
                while (locIt.hasNext()) {
                    ExternalLocation loc = locIt.next();
                    if (loc == null || loc.getLabel() == null) continue;
                    String n = loc.getLabel();
                    if (matches(n) || lib.toLowerCase().contains("user32") || lib.toLowerCase().contains("kernel32")) {
                        writeLine(out, "  " + n);
                    }
                }
            }

            writeLine(out, "");
            writeLine(out, "== Candidate Exports ==");
            it = p.getFunctionManager().getFunctions(true);
            while (it.hasNext()) {
                Function fn = it.next();
                String name = fn.getName();
                if (!name.startsWith("FUN_") && !fn.isThunk()) {
                    writeLine(out, name + " @ " + fn.getEntryPoint());
                }
            }
        }

        println("Report written: " + outFile.getAbsolutePath());
    }

    private boolean matches(String name) {
        String lower = name.toLowerCase();
        for (String t : TOKENS) {
            if (lower.contains(t)) return true;
        }
        return false;
    }

    private void writeLine(PrintWriter out, String s) {
        println(s);
        out.println(s);
    }
}

File: ghidra_scripts/PuntoHookReport.py
#@author Codex
#@category Punto
#@keybinding
#@menupath
#@toolbar

import os


def log(msg):
    print(msg)
    out.append(msg)


out = []
prog = currentProgram
fm = prog.getFunctionManager()

log("Program: %s" % prog.getName())
log("Language: %s" % str(prog.getLanguage()))
log("")

interesting_tokens = [
    "hook", "switch", "layout", "caps", "diary", "password",
    "caret", "copy", "paste", "mouse", "escape", "f12"
]

log("== Functions (interesting names) ==")
for fn in fm.getFunctions(True):
    name = fn.getName()
    lower = name.lower()
    if any(t in lower for t in interesting_tokens):
        log("%s @ %s" % (name, fn.getEntryPoint()))

log("")
log("== External Functions (imports) ==")
extmgr = prog.getExternalManager()
for lib in extmgr.getExternalLibraryNames():
    log("[%s]" % lib)
    syms = extmgr.getExternalLibrary(lib)
    if syms is None:
        continue
    for s in syms.getSymbols():
        name = s.getName()
        lower = name.lower()
        if any(t in lower for t in interesting_tokens) or lib.lower() in ["user32.dll", "kernel32.dll", "shlwapi.dll"]:
            log("  %s" % name)

log("")
log("== Exported Symbols ==")
symtab = prog.getSymbolTable()
for sym in symtab.getExternalSymbols():
    pass

# Simpler export pass: functions in EXTERNAL namespace are imports; exports usually in global namespace and referenced by entry points.
for fn in fm.getFunctions(True):
    sym = fn.getSymbol()
    if sym is None:
        continue
    if fn.isThunk():
        continue
    # In this DLL, exports are mostly near image base and Ghidra tags them with no namespace qualifiers.
    if fn.getEntryPoint().getOffset() >= prog.getImageBase().getOffset():
        name = fn.getName()
        if not name.startswith("FUN_"):
            log("%s @ %s" % (name, fn.getEntryPoint()))

log("")

report_path = os.path.join(os.getcwd(), "ghidra_pshook64_report.txt")
with open(report_path, "w") as f:
    f.write("\n".join(out))

log("Report written: %s" % report_path)
